src/main/kotlin/com/cognifide/gradle/aem/common/instance/service/repository/Node.kt

```diff
@@ -0,0 +1,210 @@
+        get() = properties.string(JcrConstants.JCR_PRIMARYTYPE)!!
+
+    /**
+     * Parent node.
+     */
+    @get:JsonIgnore
+    val parent: Node
+        get() = Node(repository, path.substringBeforeLast("/"))
+
+    /**
+     * Get all node child nodes.
+     *
+     * Because of performance issues, using method is more preferred.
+     */
+    @get:JsonIgnore
+    val children: List<Node>
+        get() = children().toList()
+
+    /**
+     * Loop over all node child nodes.
+     */
+    @Suppress("unchecked_cast")
+    fun children(): Sequence<Node> = repository.http.get("$path.harray.1.json") { asJson(it) }
+            .run {
+                try {
+                    read<JSONArray>(Property.CHILDREN.value)
+                } catch (e: PathNotFoundException) {
+                    JSONArray() // no children
+                }
+            }
+            .map { child -> child as Map<String, Any> }
+            .map { props ->
+                Node(repository, "$path/${props[Property.NAME.value]}").apply {
+                    propertiesLoaded = Properties(filterMetaProperties(props))
+                }
+            }
+            .asSequence()
+
+    /**
+     * Create or update node in repository.
+     */
+    fun save(properties: Map<String, Any?>): RepositoryResult = try {
+        repository.http.postMultipart(path, postProperties(properties) + operationProperties("")) {
+            asObjectFromJson(it, RepositoryResult::class.java)
+        }
+    } catch (e: RequestException) {
+        throw RepositoryException("Cannot save repository node: $path", e)
+    }
+
+    /**
+     * Delete node and all children from repository.
+     */
+    fun delete(): RepositoryResult = try {
+        repository.http.postMultipart(path, operationProperties("delete")) {
+            asObjectFromJson(it, RepositoryResult::class.java)
+        }
+    } catch (e: RequestException) {
+        throw RepositoryException("Cannot delete repository node: $path", e)
+    }
+
+    /**
+     * Deletes node and creates it again. Use with caution!
+     */
+    fun replace(properties: Map<String, Any?>): RepositoryResult {
+        delete()
+        return save(properties)
+    }
+
+    /**
+     * Synchronizes on demand previously loaded properties of node (by default properties are loaded lazily).
+     * Useful when saving and working on same node again (without instantiating variable).
+     */
+    fun reload() {
+        reloadProperties()
+    }
+
+    /**
+     * Search nodes by traversing a node tree.
+     * Use sequence filter method to find desired nodes.
+     */
+    fun recurse(self: Boolean = false): Sequence<Node> = sequence {
+        val stack = Stack<Node>()
+
+        if (self) {
+            stack.add(this@Node)
+        } else {
+            stack.addAll(children)
+        }
+
+        while (stack.isNotEmpty()) {
+            val current = stack.pop()
+            stack.addAll(current.children)
+            yield(current)
```

Comment:
I'm not sure if I understand this correctly but you could maybe use `yieldAll(...)` that takes lambda so when for example someone will call `Node.recursive().take(5)` it will not load all the children.

Reply:
Thx will try it